hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/pipeline/Pipeline.java

```diff
@@ -72,14 +72,63 @@ public static Codec<Pipeline> getCodec() {
   }
 
   private static final Logger LOG = LoggerFactory.getLogger(Pipeline.class);
+
+  /**
+   * Inner class that provides thread-safe access to DN list.
+   */
+  static class NodesInOrder {
+    private final List<DatanodeDetails> nodes = new ArrayList<>();
+
+    synchronized DatanodeDetails getClosest(PipelineID id, Set<DatanodeDetails> excluded) throws IOException {
+      if (nodes.isEmpty()) {
+        LOG.debug("Nodes in order is empty, delegate to getFirstNode");
```

Comment:
nit: move log to caller `getClosestNode`, similar to how log is in `getNodesInOrder`, not in `getAll`.  Mostly to avoid logging while holding the lock.

Reply:
I've added the logging in `getClosestNode`.

But I kept the `isEmpty()` check inside `getClosest` as well to facilitate debugging (because there is still a none-zero chance of TOCTOU).

Reply:
Let me clarify.  `getClosestNode` should be:

```java
    final DatanodeDetails closest = nodesInOrder.getClosest(id, excluded);
    if (closest == null) {
      LOG.debug("Nodes in order is empty, delegate to getFirstNode");
      return getFirstNode(excluded);
    }
    return closest;
```

No "between the locks" here.  Then `LOG.debug` can be removed from `getClosest`.

Reply:
ah, my bad. done